src/validator.cpp

```diff
@@ -0,0 +1,504 @@
+                    addError(err);
+                }
+                componentNames.push_back(componentName);
+            }
+            // Validate component.
+            validateComponent(component);
+        }
+    }
+    // Check for units in this model.
+    if (model->unitsCount() > 0) {
+        std::vector<std::string> unitsNames;
+        std::vector<std::string> unitsRefs;
+        std::vector<std::string> unitsImportSources;
+        for (size_t i = 0; i < model->unitsCount(); ++i) {
+            UnitsPtr units = model->getUnits(i);
+            std::string unitsName = units->getName();
+            if (unitsName.length()) {
+                if (units->isImport()) {
+                    // Check for a units_ref.
+                    std::string unitsRef = units->getImportReference();
+                    std::string importSource = units->getImport()->getSource();
+                    bool foundImportError = false;
+                    if (!unitsRef.length()) {
+                        ErrorPtr err = std::make_shared<Error>();
+                        err->setDescription("Imported units '" + unitsName +
+                                            "' does not have a valid units_ref attribute.");
+                        err->setUnits(units);
+                        err->setKind(Error::Kind::UNITS);
+                        addError(err);
+                        foundImportError = true;
+                    }
+                    // Check for a xlink:href.
+                    if (!importSource.length()) {
+                        ErrorPtr err = std::make_shared<Error>();
+                        err->setDescription("Import of units '" + unitsName +
+                                            "' does not have a valid locator xlink:href attribute.");
+                        err->setImport(units->getImport());
+                        err->setKind(Error::Kind::IMPORT);
+                        addError(err);
+                        foundImportError = true;
+                    }
+                    // Check if we already have another import from the same source with the same units_ref.
+                    // (This looks for matching enties at the same position in the source and ref vectors).
+                    if ((unitsImportSources.size() > 0) && (!foundImportError)) {
+                        if ((std::find(unitsImportSources.begin(), unitsImportSources.end(), importSource) - unitsImportSources.begin())
+                         == (std::find(unitsRefs.begin(), unitsRefs.end(), unitsRef) - unitsRefs.begin())){
+                            ErrorPtr err = std::make_shared<Error>();
+                            err->setDescription("Model '" + model->getName() +
+                                                "' contains multiple imported units from '" + importSource +
+                                                "' with the same units_ref attribute '" + unitsRef + "'.");
```

Comment:
hmmmm....ok, that has slipped in since 1.1 :) so don't worry about my comment for now...